### PR TITLE
Fixed Pagination issue

### LINKED ISSUE #9290: Sorting based on edits / page views / likes shows only 1 row on the first page of wikis
### Please describe the problem (or idea)
When we sort wikis based on `title` or `last edited`, it works fine and shows the same number of rows on the first page before and after sorting.
However, if we sort based on `edits`, `pages views` or `likes`, it shows only one row on the first page.

### Please show us where to look
https://publiclab.org/wiki

ezgif com-video-to-gif

### PR DESCRIPTION
**Pull request by gaurav2699**
Fixes #9290 (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!
image
image

### CODE REVIEW

**Comment by gaurav2699** [comment]
@cesswairimu @jywarren please review

File: app/controllers/wiki_controller.rb [hunk 1]
```diff
@@ -341,7 +341,7 @@ def index
       .references(:node_revisions)
       .group('node_revisions.nid, node_revisions.vid')
       .order(order_string)
-      .where("node_revisions.status = 1 AND node.status = 1 AND type = 'page'"))
+      .where("node_revisions.status = 1 AND node.status = 1 AND (type = 'page' OR type = 'tool' OR type = 'place')"), items: 10)
```
Comment (RuthNjeri) [inline app/controllers/wiki_controller.rb:344]: Hi @gaurav2699, thanks for opening this up.

I know that `type = 'page'` represents wikis, and I think that this query before was meant to display wikis only, why do we need to add `type = tool OR type = place`? Are they related to the wiki pages?

File: app/controllers/wiki_controller.rb [hunk 2]
```diff
@@ -353,7 +353,7 @@ def stale
       .references(:node_revisions)
       .group('node_revisions.nid, node_revisions.vid')
       .order('node_revisions.timestamp ASC')
-      .where("node_revisions.status = 1 AND node.status = 1 AND type = 'page'"))
+      .where("node_revisions.status = 1 AND node.status = 1 AND (type = 'page' OR type = 'tool' OR type = 'place')"), items: 10)
```
Comment (RuthNjeri) [inline app/controllers/wiki_controller.rb:356]: See comment above on `type = tool OR type = place`

**Comment by RuthNjeri** [comment]
Hi, @gaurav2699 also noting this issue https://github.com/publiclab/plots2/issues/714 that aims to remove all references to `tool and place`, which could render your code obsolete

File: app/controllers/wiki_controller.rb (hunk 2 above)
Comment (gaurav2699) [reply app/controllers/wiki_controller.rb]: hey @RuthNjeri, yeah I think my version of code was old and hence github got confused. I only added the "items: 10" part. I will make the required changes. Thank you so much

Comment (gaurav2699) [reply app/controllers/wiki_controller.rb]: @RuthNjeri You can check now

**Comment by gaurav2699** [comment]
@RuthNjeri You can check now. Thanks!

**Review by Tlazypanda — COMMENTED** [review]
@gaurav2699 LGTM :tada: can we also write a functional test to make sure this doesn't break later?

**Comment by gaurav2699** [comment]
@Tlazypanda I added a functional test, please review and check if it's alright! Thanks a lot!

File: test/functional/wiki_controller_test.rb [hunk 3]
```diff
@@ -31,6 +31,21 @@ def teardown
...
+  test 'should paginate the wikis' do
+    31.times{
+    post :create,
+         params: {
+           uid:   users(:bob).id,
+           title: 'Test',
+           body:  'This is fascinating documentation about balloon mapping.',
+           tags:  'balloon-mapping'
+         }
+    }
```
Comment (RuthNjeri) [inline test/functional/wiki_controller_test.rb:43]: Thanks for creating the test @gaurav2699, I am curious if this request might take a long time to run? 
There is a file where the test database can be preloaded with data, `test/fixtures` and within it, is the `nodes.yml` file  `test/fixtures/nodes.yml` that contains wikis as well as research notes, to create a wiki within the `nodes.yml` file the `type: "page"`.

Comment (gaurav2699) [reply test/functional/wiki_controller_test.rb:43]: hey @RuthNjeri, thanks for the review. I am aware of nodes.yml but it contained only 10 nodes that are wikis(nodes with type: 'page'). To test if the pagination would work, having just 10 nodes was too less as every page will contain 10 items. Hence I created a few wikis, though I agree 31 is a big number, I will change it to a lesser value.

Comment (RuthNjeri) [reply test/functional/wiki_controller_test.rb]: I did not mean that the number is huge, just that a POST request being made 31 times seems like an expensive call as opposed to having prepopulated nodes for this...what do you think?

Comment (gaurav2699) [reply test/functional/wiki_controller_test.rb]: Or should I create few more wikis in the nodes.yml file only, but I think a few other tests depend on the wikis in that hence I didn't feel it was a good idea to touch that.

Comment (gaurav2699) [reply test/functional/wiki_controller_test.rb]: Anyways, I reduced 31 to 6, is that good enough?

Comment (gaurav2699) [reply test/functional/wiki_controller_test.rb]: Anyways, I reduced 31 to 12, is that good enough?

Comment (gaurav2699) [reply test/functional/wiki_controller_test.rb]: Yeah, I agree that this is slower. So should I add few more wikis in the node.yml file itself?

Comment (RuthNjeri) [reply test/functional/wiki_controller_test.rb]: Hi @gaurav2699 I think it's okay for now since the number has reduced to 12...

**Comment by cesswairimu** [comment]
Hi @gaurav2699 , there are one test that was added on main that is failing because of the new wikis created in your test block...would you like to help make these a little more dynamic this would involve changing [this test](https://github.com/publiclab/plots2/blob/main/test/functional/wiki_controller_test.rb#L34-L39) to smth like 

```
test 'should sort by last edited' do
   get :index, params: { sort: "last_edited" }
   wikis = assigns(:wikis)
   wiki1 = wikis.order(changed: :desc).first
   wiki2 = wikis.order(changed: :desc).reverse.first
   assert_equal(wikis.first.title, wiki1.title)
   assert_equal(wikis.last.title, wiki2.title)
end
```
what do you think?  Thanks

**Comment by gaurav2699** [comment]
Thanks for the guidance @cesswairimu. I made the change and now the functional tests doesn't fail. Please review. Thanks a lot.

**Review by cesswairimu — APPROVED** [review]
Great great job here @gaurav2699 thanks a lot

**Pull request merged**